src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_advection.jl

```diff
@@ -0,0 +1,38 @@
+using Oceananigans.Advection: div_Uu, div_Uv
+using Oceananigans.Operators
+
+######
+###### Horizontally-vector-invariant formulation of momentum advection
+######
+
+@inline Δy_vᶜᶠᶜ(i, j, k, grid, v) = @inbounds Δy(i, j, k, grid) * v[i, j, k]
+@inline Δx_uᶠᶜᶜ(i, j, k, grid, u) = @inbounds Δx(i, j, k, grid) * u[i, j, k]
+
+@inline Axyᶠᶠᵃ(i, j, k, grid) = Δx(i, j, k, grid) * Δy(i, j, k, grid)
+
+@inline Γᶠᶠᶜ(i, j, k, grid, u, v) = δxᶠᵃᵃ(i, j, k, grid, Δy_vᶜᶠᶜ, v) - δyᵃᶠᵃ(i, j, k, grid, Δx_uᶠᶜᶜ, u)
+@inline ζ₃ᶠᶠᶜ(i, j, k, grid, u, v) = Γᶠᶠᶜ(i, j, k, grid, u, v) / Axyᶠᶠᵃ(i, j, k, grid)
+
+@inline ϕ²(i, j, k, grid, ϕ) = @inbounds ϕ[i, j, k]^2
+@inline Khᶜᶜᶜ(i, j, k, grid, u, v) = (ℑxᶜᵃᵃ(i, j, k, grid, ϕ², u) + ℑyᵃᶜᵃ(i, j, k, grid, ϕ², v)) / 2
+
+@inbounds ζ₂wᶠᶜᶠ(i, j, k, grid, u, w) = ℑxᶠᵃᵃ(i, j, k, grid, w) * δzᵃᵃᶠ(i, j, k, grid, u) / ΔzC(i, j, k, grid)
+@inbounds ζ₁wᶜᶠᶠ(i, j, k, grid, v, w) = ℑyᵃᶠᵃ(i, j, k, grid, w) * δzᵃᵃᶠ(i, j, k, grid, v) / ΔzC(i, j, k, grid)
+
+@inline U_dot_∇u(i, j, k, grid, advection::VectorInvariant, U) = (
```

Comment:
I don't recognize this form of the equations, but I presume you have checked them multiple times and the results look pretty reasonable.

Reply:
It can be derived from Naiver-Stokes using the identity

<img width="266" alt="image" src="https://user-images.githubusercontent.com/15271942/108212646-21702800-70fc-11eb-883a-8e1eaae1d22f.png">

I am following https://mitgcm.readthedocs.io/en/latest/algorithm/algorithm.html#vector-invariant-momentum-equations for the finite volume implementation, though I'm not 100% sure my implementation will work on a curvilinear grid (future PR).

Reply:
Ah, thanks for the reference.  Very helpful.